engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -40,366 +37,170 @@
-      boolean self,
-      Object other,
-      @Shared("interop") @CachedLibrary(limit = "10") InteropLibrary interop) {
-    try {
-      return self == interop.asBoolean(other);
-    } catch (UnsupportedMessageException ex) {
-      return false;
-    }
+  @NeverDefault
+  static EqualsNode build() {
+    return create();
   }
 
-  @Specialization
-  boolean equalsLongLong(long self, long other) {
-    return self == other;
+  @NeverDefault
+  public static EqualsNode create() {
+    return new EqualsNode(EqualsSimpleNode.build(), TypeOfNode.build());
   }
 
-  @Specialization
-  boolean equalsLongBool(long self, boolean other) {
-    return false;
-  }
-
-  @Specialization
-  boolean equalsLongDouble(long self, double other) {
-    return (double) self == other;
-  }
-
-  @Specialization
-  boolean equalsLongText(long self, Text other) {
-    return false;
-  }
-
-  @Specialization
-  @TruffleBoundary
-  boolean equalsLongBigInt(long self, EnsoBigInteger other) {
-    if (BigIntegerOps.fitsInLong(other.getValue())) {
-      return BigInteger.valueOf(self).compareTo(other.getValue()) == 0;
-    } else {
-      return false;
-    }
-  }
-
-  @Specialization
-  boolean equalsLongInterop(
-      long self,
-      Object other,
-      @Shared("interop") @CachedLibrary(limit = "10") InteropLibrary interop) {
-    try {
-      return self == interop.asLong(other);
-    } catch (UnsupportedMessageException ex) {
-      return false;
-    }
+  @NeverDefault
+  public static EqualsNode getUncached() {
+    return new EqualsNode(EqualsSimpleNodeGen.getUncached(), TypeOfNode.getUncached());
   }
 
-  @Specialization
-  boolean equalsDoubleDouble(double self, double other) {
-    if (Double.isNaN(self) || Double.isNaN(other)) {
-      return false;
-    } else {
-      return self == other;
-    }
-  }
-
-  @Specialization
-  boolean equalsDoubleLong(double self, long other) {
-    return self == (double) other;
-  }
-
-  @Specialization
-  boolean equalsDoubleBool(double self, boolean other) {
-    return false;
-  }
-
-  @Specialization
-  boolean equalsDoubleText(double self, Text other) {
-    return false;
-  }
-
-  @Specialization
-  boolean equalsDoubleInterop(
-      double self,
-      Object other,
-      @Shared("interop") @CachedLibrary(limit = "10") InteropLibrary interop) {
-    try {
-      if (interop.fitsInDouble(other)) {
-        return self == interop.asDouble(other);
+  /**
+   * Compares two objects for equality. If the {@link EqualsSimpleNode simple check} fails, it tries
+   * to convert first argument to the second one and compare again.
+   *
+   * @param frame the stack frame we are executing at
+   * @param self the self object
+   * @param other the other object
+   * @return {@code true} if {@code self} and {@code that} seem equal
+   */
+  public boolean execute(
+      VirtualFrame frame, @AcceptsError Object self, @AcceptsError Object other) {
+    var areEqual = node.execute(frame, self, other);
+    if (!areEqual) {
```

Comment:
@Akirathan, I have renamed original `EqualsNode` to `EqualsSimpleNode`. Then I placed this new `EqualsNode` in front of the original one and in case the original one _reports inequality_ I try harder with conversions via `WithConversionNode`.

Functionally it should be fine. The effect on benchmarks needs yet to be verified.